<a href="https://colab.research.google.com/github/ritwiks9635/My_Neural_Network_Architecture/blob/main/Classification_with_Gated_Residual_and_Variable_Selection_Networks.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Classification with Gated Residual and Variable Selection Networks**

## Introduction

This example demonstrates the use of Gated
Residual Networks (GRN) and Variable Selection Networks (VSN), proposed by
Bryan Lim et al. in
[Temporal Fusion Transformers (TFT) for Interpretable Multi-horizon Time Series Forecasting](https://arxiv.org/abs/1912.09363),
for structured data classification. GRNs give the flexibility to the model to apply
non-linear processing only where needed. VSNs allow the model to softly remove any
unnecessary noisy inputs which could negatively impact performance.
Together, those techniques help improving the learning capacity of deep neural
network models.

Note that this example implements only the GRN and VSN components described in
in the paper, rather than the whole TFT model, as GRN and VSN can be useful on
their own for structured data learning tasks.


To run the code you need to use TensorFlow 2.3 or higher.

In [3]:
import math
import numpy as np
import pandas as pd
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

In [4]:
data = pd.read_csv("adult.csv")
test_data = pd.read_csv("adult.test.csv", header = None, names = list(data.columns))
data.head()

Age          Workclass  Final Weight   Education  EducationNum  \
0   39          State-gov         77516   Bachelors            13   
1   50   Self-emp-not-inc         83311   Bachelors            13   
2   38            Private        215646     HS-grad             9   
3   53            Private        234721        11th             7   
4   28            Private        338409   Bachelors            13   

        Marital Status          Occupation    Relationship    Race   Gender  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   Capital Gain  capital loss  Hours per Week  Native Country  Income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [5]:
test_data.head()

Age   Workclass  Final Weight      Education  EducationNum  \
0   25     Private        226802           11th             7   
1   38     Private         89814        HS-grad             9   
2   28   Local-gov        336951     Assoc-acdm            12   
3   44     Private        160323   Some-college            10   
4   18           ?        103497   Some-college            10   

        Marital Status          Occupation Relationship    Race   Gender  \
0        Never-married   Machine-op-inspct    Own-child   Black     Male   
1   Married-civ-spouse     Farming-fishing      Husband   White     Male   
2   Married-civ-spouse     Protective-serv      Husband   White     Male   
3   Married-civ-spouse   Machine-op-inspct      Husband   Black     Male   
4        Never-married                   ?    Own-child   White   Female   

   Capital Gain  capital loss  Hours per Week  Native Country   Income  
0             0             0              40   United-States   <=50K.  
1             0             0              50   United-States   <=50K.  
2             0             0              40   United-States    >50K.  
3          7688             0              40   United-States    >50K.  
4             0             0              30   United-States   <=50K.

In [6]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
data["Income"] = le.fit_transform(data["Income"])
test_data["Income"] = le.fit_transform(test_data["Income"])

In [7]:
random_selection = np.random.rand(len(data.index)) <= 0.85
train_data = data[random_selection]
valid_data = data[~random_selection]

In [8]:
train_data_file = "train_data.csv"
valid_data_file = "valid_data.csv"
test_data_file = "test_data.csv"

train_data.to_csv(train_data_file, index=False, header=False)
valid_data.to_csv(valid_data_file, index=False, header=False)
test_data.to_csv(test_data_file, index=False, header=False)

In [9]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 27668 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Age             27668 non-null  int64 
 1   Workclass       27668 non-null  object
 2   Final Weight    27668 non-null  int64 
 3   Education       27668 non-null  object
 4   EducationNum    27668 non-null  int64 
 5   Marital Status  27668 non-null  object
 6   Occupation      27668 non-null  object
 7   Relationship    27668 non-null  object
 8   Race            27668 non-null  object
 9   Gender          27668 non-null  object
 10  Capital Gain    27668 non-null  int64 
 11  capital loss    27668 non-null  int64 
 12  Hours per Week  27668 non-null  int64 
 13  Native Country  27668 non-null  object
 14  Income          27668 non-null  int64 
dtypes: int64(7), object(8)
memory usage: 3.4+ MB


In [44]:
categorical_colunms = [cat for cat in data.columns if data[cat].dtype == "object"]
numerical_columns = [num for num in data.columns if data[num].dtype != "object"]
print(categorical_colunms)
print(numerical_columns)

['Workclass', 'Education', 'Marital Status', 'Occupation', 'Relationship', 'Race', 'Gender', 'Native Country']
['Age', 'Final Weight', 'EducationNum', 'Capital Gain', 'capital loss', 'Hours per Week', 'Income']


In [ ]:
CSV_HEADER = list(data.columns)

## Define dataset metadata

Here, we define the metadata of the dataset that will be useful for reading and
parsing the data into input features, and encoding the input features with respect
to their types.

In [35]:
# Target feature name.
TARGET_FEATURE_NAME = "Income"

# Numeric feature names.
NUMERIC_FEATURE_NAMES = [
'Age',
 'Final Weight',
 'EducationNum',
 'Capital Gain',
 'capital loss',
 'Hours per Week'
]# Categorical features and their vocabulary lists.
# Note that we add 'v=' as a prefix to all categorical feature values to make
# sure that they are treated as strings.
CATEGORICAL_FEATURES_WITH_VOCABULARY = {
    feature_name: sorted([str(value) for value in list(data[feature_name].unique())])
    for feature_name in CSV_HEADER
    if feature_name
    not in list(NUMERIC_FEATURE_NAMES + [TARGET_FEATURE_NAME])
}
# All features names.
FEATURE_NAMES = NUMERIC_FEATURE_NAMES + list(
    CATEGORICAL_FEATURES_WITH_VOCABULARY.keys()
)
# Feature default values.
COLUMN_DEFAULTS = [
    [0.0]
    if feature_name in NUMERIC_FEATURE_NAMES + [TARGET_FEATURE_NAME]
    else ["NA"]
    for feature_name in CSV_HEADER
]

## Create a `tf.data.Dataset` for training and evaluation

We create an input function to read and parse the file, and convert features and
labels into a [`tf.data.Dataset`](https://www.tensorflow.org/guide/datasets) for
training and evaluation.

In [36]:
from tensorflow.keras.layers import StringLookup


def process(features, target):
    for feature_name in features:
        if feature_name in CATEGORICAL_FEATURES_WITH_VOCABULARY:
            # Cast categorical feature values to string.
            features[feature_name] = tf.cast(features[feature_name], tf.dtypes.string)
    # Get the instance weight.
   # weight = features.pop(WEIGHT_COLUMN_NAME)
    return features, target


def get_dataset_from_csv(csv_file_path, shuffle=False, batch_size=128):

    dataset = tf.data.experimental.make_csv_dataset(
        csv_file_path,
        batch_size=batch_size,
        column_names=CSV_HEADER,
        column_defaults=COLUMN_DEFAULTS,
        label_name=TARGET_FEATURE_NAME,
        num_epochs=1,
        header=False,
        shuffle=shuffle,
    ).map(process)

    return dataset

## Create model inputs

In [37]:
def create_model_inputs():
    inputs = {}
    for feature_name in FEATURE_NAMES:
        if feature_name in NUMERIC_FEATURE_NAMES:
            inputs[feature_name] = layers.Input(
                name=feature_name, shape=(), dtype=tf.float32
            )
        else:
            inputs[feature_name] = layers.Input(
                name=feature_name, shape=(), dtype=tf.string
            )
    return inputs

## Encode input features

For categorical features, we encode them using `layers.Embedding` using the
`encoding_size` as the embedding dimensions. For the numerical features,
we apply linear transformation using `layers.Dense` to project each feature into
`encoding_size`-dimensional vector. Thus, all the encoded features will have the
same dimensionality.

In [38]:
def encode_inputs(inputs, encoding_size):
    encoded_features = []
    for feature_name in inputs:
        if feature_name in CATEGORICAL_FEATURES_WITH_VOCABULARY:
            vocabulary = CATEGORICAL_FEATURES_WITH_VOCABULARY[feature_name]
            # Create a lookup to convert a string values to an integer indices.
            # Since we are not using a mask token nor expecting any out of vocabulary
            # (oov) token, we set mask_token to None and  num_oov_indices to 0.
            index = StringLookup(
                vocabulary=vocabulary, mask_token=None, num_oov_indices=0
            )
            # Convert the string input values into integer indices.
            value_index = index(inputs[feature_name])
            # Create an embedding layer with the specified dimensions
            embedding_ecoder = layers.Embedding(
                input_dim=len(vocabulary), output_dim=encoding_size
            )
            # Convert the index values to embedding representations.
            encoded_feature = embedding_ecoder(value_index)
        else:
            # Project the numeric feature to encoding_size using linear transformation.
            encoded_feature = tf.expand_dims(inputs[feature_name], -1)
            encoded_feature = layers.Dense(units=encoding_size)(encoded_feature)
        encoded_features.append(encoded_feature)
    return encoded_features

## Implement the Gated Linear Unit

[Gated Linear Units (GLUs)](https://arxiv.org/abs/1612.08083) provide the
flexibility to suppress input that are not relevant for a given task.

In [39]:
class GatedLinearUnit(layers.Layer):
    def __init__(self, units):
        super().__init__()
        self.linear = layers.Dense(units)
        self.sigmoid = layers.Dense(units, activation="sigmoid")

    def call(self, inputs):
        return self.linear(inputs) * self.sigmoid(inputs)

## Implement the Gated Residual Network

The Gated Residual Network (GRN) works as follows:

1. Applies the nonlinear ELU transformation to the inputs.
2. Applies linear transformation followed by dropout.
4. Applies GLU and adds the original inputs to the output of the GLU to perform skip
(residual) connection.
6. Applies layer normalization and produces the output.

In [40]:
class GatedResidualNetwork(layers.Layer):
    def __init__(self, units, dropout_rate):
        super().__init__()
        self.units = units
        self.elu_dense = layers.Dense(units, activation="elu")
        self.linear_dense = layers.Dense(units)
        self.dropout = layers.Dropout(dropout_rate)
        self.gated_linear_unit = GatedLinearUnit(units)
        self.layer_norm = layers.LayerNormalization()
        self.project = layers.Dense(units)

    def call(self, inputs):
        x = self.elu_dense(inputs)
        x = self.linear_dense(x)
        x = self.dropout(x)
        if inputs.shape[-1] != self.units:
            inputs = self.project(inputs)
        x = inputs + self.gated_linear_unit(x)
        x = self.layer_norm(x)
        return x

## Implement the Variable Selection Network

The Variable Selection Network (VSN) works as follows:

1. Applies a GRN to each feature individually.
2. Applies a GRN on the concatenation of all the features, followed by a softmax to
produce feature weights.
3. Produces a weighted sum of the output of the individual GRN.

Note that the output of the VSN is [batch_size, encoding_size], regardless of the
number of the input features.

In [41]:
class VariableSelection(layers.Layer):
    def __init__(self, num_features, units, dropout_rate):
        super().__init__()
        self.grns = list()
        # Create a GRN for each feature independently
        for idx in range(num_features):
            grn = GatedResidualNetwork(units, dropout_rate)
            self.grns.append(grn)
        # Create a GRN for the concatenation of all the features
        self.grn_concat = GatedResidualNetwork(units, dropout_rate)
        self.softmax = layers.Dense(units=num_features, activation="softmax")

    def call(self, inputs):
        v = layers.concatenate(inputs)
        v = self.grn_concat(v)
        v = tf.expand_dims(self.softmax(v), axis=-1)

        x = []
        for idx, input in enumerate(inputs):
            x.append(self.grns[idx](input))
        x = tf.stack(x, axis=1)

        outputs = tf.squeeze(tf.matmul(v, x, transpose_a=True), axis=1)
        return outputs

## Create Gated Residual and Variable Selection Networks model

In [42]:
def create_model(encoding_size):
    inputs = create_model_inputs()
    feature_list = encode_inputs(inputs, encoding_size)
    num_features = len(feature_list)

    features = VariableSelection(num_features, encoding_size, dropout_rate)(
        feature_list
    )

    outputs = layers.Dense(units=1, activation="sigmoid")(features)
    model = keras.Model(inputs=inputs, outputs=outputs)
    return model

In [43]:
learning_rate = 0.001
dropout_rate = 0.15
batch_size = 265
num_epochs = 20
encoding_size = 16

model = create_model(encoding_size)
model.compile(
    optimizer=keras.optimizers.Adam(learning_rate=learning_rate),
    loss=keras.losses.BinaryCrossentropy(),
    metrics=[keras.metrics.BinaryAccuracy(name="accuracy")],
)


# Create an early stopping callback.
early_stopping = tf.keras.callbacks.EarlyStopping(
    monitor="val_loss", patience=5, restore_best_weights=True
)

print("Start training the model...")
train_dataset = get_dataset_from_csv(
    train_data_file, shuffle=True, batch_size=batch_size
)
valid_dataset = get_dataset_from_csv(valid_data_file, batch_size=batch_size)
model.fit(
    train_dataset,
    epochs=num_epochs,
    validation_data=valid_dataset,
    callbacks=[early_stopping],
)
print("Model training finished.")

print("Evaluating model performance...")
test_dataset = get_dataset_from_csv(test_data_file, batch_size=batch_size)
_, accuracy = model.evaluate(test_dataset)
print(f"Test accuracy: {round(accuracy * 100, 2)}%")

Start training the model...
Epoch 1/20
105/105 [==============================] - 25s 63ms/step - loss: 0.4253 - accuracy: 0.7898 - val_loss: 0.3758 - val_accuracy: 0.8265
Epoch 2/20
105/105 [==============================] - 4s 31ms/step - loss: 0.3529 - accuracy: 0.8373 - val_loss: 0.3678 - val_accuracy: 0.8277
Epoch 3/20
105/105 [==============================] - 4s 37ms/step - loss: 0.3440 - accuracy: 0.8422 - val_loss: 0.3591 - val_accuracy: 0.8312
Epoch 4/20
105/105 [==============================] - 5s 40ms/step - loss: 0.3401 - accuracy: 0.8427 - val_loss: 0.3582 - val_accuracy: 0.8318
Epoch 5/20
105/105 [==============================] - 4s 31ms/step - loss: 0.3376 - accuracy: 0.8435 - val_loss: 0.3560 - val_accuracy: 0.8347
Epoch 6/20
105/105 [==============================] - 5s 43ms/step - loss: 0.3369 - accuracy: 0.8443 - val_loss: 0.3565 - val_accuracy: 0.8343
Epoch 7/20
105/105 [==============================] - 4s 31ms/step - loss: 0.3349 - accuracy: 0.8450 - val_loss: 

You should achieve more than 85% accuracy on the test set.

To increase the learning capacity of the model, you can try increasing the
`encoding_size` value, or stacking multiple GRN layers on top of the VSN layer.
This may require to also increase the `dropout_rate` value to avoid overfitting.